### DISCUSSION TITLE
Clarification on Many-to-Many API Rule

### DISCUSSION THREAD

**Discussion by uxsamok** [pocketbase/pocketbase#4040] [Q&A]
Why does this many to many API Rule work and should I rely on this?

**Short version** 
Why does this work for checking if any item of Array2 is present as an item in Array1?
`Array1 ?~ Array2`

**TLDR**
I have three collections:
- ORGS with a list of organisations
- USERS with an org relation field for multiple orgs.
- ASSETS with an org relation field for multiple orgs.

My goal is to allow a USER to access all ASSETS where there's at least one ORG in common.

The API Rule I used for the ASSETS collection is
`org ?~ @request.auth.org.id`
Surprisingly for me it worked but I don't understand why and can I safely rely on that?

Any help would be greatly appreciated. 
I know "why did it work" is an unconventional question :)

**Comment by ganigeorgiev** [comment]
Array values are stored as serialized JSON array, eg. `["id1", "id2"]`.
The `~`/`?~` operator is converted to a `x LIKE %y%` SQL expression, aka. it only checks whether "x" contains the "y" string.

> should I rely on this

It depends. If you want an exact match use `?=` instead.

_Edit: note that with v0.20.1 you can also start the executable with `--dev` and explore the generated SQL queries in the console._

**Reply by uxsamok** [reply]
Thank you very much for the reply! Just to double check I understand this. 

The expression `["IDa", "IDb"] ?~ ["ID1", "ID2"]` checks if at least one item of the right array is present in the left array, correct? 

That is exactly the functionality I needed, and it works perfectly. I just did not expect that `?~` compares each item on both sides of the operator. So I worried I was using it in a wrong way.

**Reply by ganigeorgiev** [reply]
No. Your example doesn't express the same thing as in your initial post.
In the above you are comparing the raw `json` field values, but in your initial post you are using dot notation to target a specific relation field (in your case the id).

Again, if you want an exact id match, use `?=`.

**Reply by uxsamok** [reply]
Thanks again, but I think I'm still confused.

In my case, in the USER collection, org in a relation field pointing at **multiple org** in the ORGS collection. 
Each user related to multiple organisations. I thought that means that `@request.auth.org.id` returns an array of ids for multiple organisations the authorised user is related to.

So `org` field is an array and `@request.auth.org.id` is an array too. Is that the case?
I tried and `?~` did work and `?=` did not work (returned empty result) in my case.

**Reply by ganigeorgiev** [reply]
Sorry, I don't know how to explain it better. `~` is a LIKE expression.

Again, if you are targeting a specific field and want exact match, use `?=`:
```
org.id ?= @request.auth.org.id
```

Add the `--dev` flag when starting the executable to inspect the constructed SQL queries.

**Reply by uxsamok** [reply]
Thank you very much for your help! And I love PocketBase :)
